### PR TITLE
robustness engine updates for concurrency

### PR DESCRIPTION
**Pull request by stevejoachim**
The multiclient robustness tests will have multiple clients calling `Engine.ExecAction` concurrently. The goal of this PR is to update the engine `Log` and `Stats` structs so that they can be safely accessed and modified from `Engine.ExecAction` concurrently.

### CODE REVIEW

File: tests/robustness/engine/action.go [hunk 1]
```diff
@@ -22,9 +22,16 @@ func (e *Engine) ExecAction(ctx context.Context, actionKey ActionKey, opts map[s
...
+	e.RunStats.IncrementActionCount()
+	e.CumulativeStats.IncrementActionCount()
+	log.Printf(
+		"Engine executing ACTION: name=%q actionCount=%v totActCount=%v t=%vs (%vs)",
+		actionKey,
+		e.RunStats.GetActionCount(),
+		e.CumulativeStats.GetActionCount(),
+		e.RunStats.getLifetimeSeconds(),
+		e.getRuntimeSeconds(),
+	)
```
Comment (carlbraganza) [inline tests/robustness/engine/action.go:34]: All this should be atomic

File: tests/robustness/engine/action.go [hunk 2]
```diff
@@ -53,25 +60,17 @@ func (e *Engine) ExecAction(ctx context.Context, actionKey ActionKey, opts map[s
 	// If error was just a no-op, don't bother logging the action
 	switch {
 	case errors.Is(err, robustness.ErrNoOp):
-		e.RunStats.NoOpCount++
-		e.CumulativeStats.NoOpCount++
+		e.RunStats.IncrementNoOpCount()
+		e.CumulativeStats.IncrementNoOpCount()
```
Comment (carlbraganza) [inline tests/robustness/engine/action.go:64]: Needs to be atomic

File: tests/robustness/engine/action.go (hunk 1 above)
Comment (carlbraganza) [reply tests/robustness/engine/action.go:34]: Consider calling `e.IncrementActionCount()` instead, where
```
func (e *Engine) IncrementActionCount() {
    e.statsMux.Lock()
    defer e.statsMux.Unlock()
    
    // all the critical section logic here
}
```

Comment (carlbraganza) [reply tests/robustness/engine/action.go:34]: Likewise for the decrement logic

File: tests/robustness/engine/action.go [hunk 3]
```diff
@@ -60,23 +52,60 @@ func (e *Engine) ExecAction(ctx context.Context, actionKey ActionKey, opts map[s
 	// If error was just a no-op, don't bother logging the action
 	switch {
 	case errors.Is(err, robustness.ErrNoOp):
-		e.RunStats.IncrementNoOpCount()
-		e.CumulativeStats.IncrementNoOpCount()
+		e.RunStats.NoOpCount++
+		e.CumulativeStats.NoOpCount++
```
Comment (carlbraganza) [inline tests/robustness/engine/action.go]: Got to protect these with the `statsMux`

File: tests/robustness/engine/action.go [hunk 4]
```diff
@@ -60,23 +52,60 @@ func (e *Engine) ExecAction(ctx context.Context, actionKey ActionKey, opts map[s
...
-	e.CumulativeStats.RecordAction(actionKey, st, err)
-
-	e.EngineLog.AddCompleted(logEntry, err)
+	e.updatePerActionStats(actionKey, st, err)
+	e.logCompleted(logEntry, err)
 
 	return out, err
 }
 
+func (e *Engine) incrActionCountAndLog(actionKey ActionKey) {
```
Comment (carlbraganza) [inline tests/robustness/engine/action.go:70]: Both these functions should probably be moved to the `stats.go` file. Perhaps also use "stats" as a prefix in the name.

File: tests/robustness/engine/action.go [hunk 5]
```diff
@@ -111,8 +140,8 @@ func (e *Engine) checkErrRecovery(ctx context.Context, incomingErr error, action
 			return outgoingErr
 		}
 
-		e.RunStats.IncrementDataPurgeCount()
-		e.CumulativeStats.IncrementDataPurgeCount()
+		e.RunStats.DataPurgeCount++
+		e.CumulativeStats.DataPurgeCount++
```
Comment (carlbraganza) [inline tests/robustness/engine/action.go]: Suggestion: Put these stats related counter updates into subroutine which accepts some sort of discriminator. Then one can safely protect them under all circumstances. e.g.
```
func (e *Engine) statsUpdateCounters(updateType string) {
	e.statsMux.Lock()
	defer e.statsMux.Unlock()
	switch updateType {
         case "incr-no-op": ...
         case "incr-data-purge":  ...
         case "incr-data-restore": ... 
         case "incr-error-recovery": ...
        }
}
```

**Review by carlbraganza — CHANGES_REQUESTED** [review]
minor changes

**Comment by stevejoachim** [comment]
@carlbraganza Thank you for catching those issues and suggesting a better way to structure everything.

I made some changes based on your feedback. Please let me know what you think.

File: tests/robustness/engine/stats.go [hunk 6]
```diff
@@ -155,3 +156,65 @@ func (e *Engine) getTimestampS() int64 {
...
 	return durationToSec(e.CumulativeStats.RunTime + clock.Since(e.RunStats.CreationTime))
 }
+
+type statsUpdatetype int
+
+const (
+	incrNoOp statsUpdatetype = iota
+	incrDataPurge
+	incrDataRestore
+	incrErrorRecovery
```
Comment (carlbraganza) [inline tests/robustness/engine/stats.go:166]: Could have used "stats" as a prefix here too :-)

File: tests/robustness/engine/action.go [hunk 7]
```diff
@@ -53,29 +51,25 @@ func (e *Engine) ExecAction(ctx context.Context, actionKey ActionKey, opts map[s
...
 
-	if e.CumulativeStats.PerActionStats != nil && e.CumulativeStats.PerActionStats[actionKey] == nil {
-		e.CumulativeStats.PerActionStats[actionKey] = new(ActionStats)
-	}
+	return out, err
+}
 
-	e.RunStats.PerActionStats[actionKey].Record(st, err)
-	e.CumulativeStats.PerActionStats[actionKey].Record(st, err)
+func (e *Engine) logCompleted(logEntry *LogEntry, err error) {
```
Comment (carlbraganza) [inline tests/robustness/engine/action.go:68]: This function should be moved to `log.go`

**Review by carlbraganza — APPROVED** [review]

**Comment by stevejoachim** [comment]
Thanks, I made the changes you suggested.

**Pull request merged**